src/testing/tree_construction.rs

```diff
@@ -50,188 +32,80 @@ pub struct Test {
     pub line: usize,
     /// The specification of the test provided in the test file
     pub spec: TestSpec,
-    /// The document tree that is expected to be parsed
+    /// The document tree as found in the spec converted to an array
     pub document: Vec<String>,
 }
 
-/// Holds the result of a single "node" (which is either an element, text or comment)
-pub enum NodeResult {
-    /// An attribute of an element node did not match
-    AttributeMatchFailure {
-        name: String,
-        actual: String,
-        expected: String,
-    },
-
-    /// The actual element does not match the expected element
-    ElementMatchFailure {
-        name: String,
-        actual: String,
-        expected: String,
-    },
-
-    /// The element matches the expected element
-    ElementMatchSuccess {
-        actual: String,
-    },
-
-    /// A text node did not match
-    TextMatchFailure {
-        actual: String,
-        expected: String,
-        text: String,
-    },
-
-    // A doctype node did not match
-    DocTypeMatchFailure {
-        actual: String,
-        expected: String,
-    },
-
-    /// A comment node did not match
-    CommentMatchFailure {
-        actual: String,
-        expected: String,
-        comment: String,
-    },
-
-    /// A text node matches
-    TextMatchSuccess {
-        expected: String,
-    },
-}
-
-pub struct SubtreeResult {
-    pub node: Option<NodeResult>,
-    pub children: Vec<SubtreeResult>,
-    next_expected_idx: Option<usize>,
-}
-
-impl SubtreeResult {
-    pub fn valid(&self) -> bool {
-        self.next_expected_idx.is_some()
-    }
-}
-
-#[derive(PartialEq)]
-pub enum ErrorResult {
-    /// Found the correct error
-    Success { actual: TestError },
-    /// Didn't find the error (not even with incorrect position)
-    Failure {
-        actual: TestError,
-        expected: TestError,
-    },
-    /// Found the error, but on an incorrect position
-    PositionFailure {
-        actual: TestError,
-        expected: TestError,
-    },
-}
-
-pub struct TestResult {
-    pub root: SubtreeResult,
-    pub actual_document: DocumentHandle,
-    pub actual_errors: Vec<ParseError>,
-}
-
-impl TestResult {
-    pub fn success(&self) -> bool {
-        self.root.valid()
+impl Clone for Test {
+    fn clone(&self) -> Self {
+        Self {
+            file_path: self.file_path.clone(),
+            line: self.line,
+            spec: self.spec.clone(),
+            document: self.document.clone(),
+        }
```

Comment:
Wondering whether we need an explicit `Clone` implementation, instead of just doing
```rust
#[derive(Copy, Clone, Debug, ...)]
pub struct Test {
  // ...
```

Reply:
I think we can now after a refactor that i did and didn't remove this clone